### PR TITLE
[wip] fault-tolerant random state 

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Continuation of #8366

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [ ] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -117,13 +136,87 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
 
+    def __getitem__(self, item):
+        if self._cached_state_dict is not None:
+            self._set_rng_states(self._cached_state_dict[self.worker_id]["rng_states"])
+            self._cached_state_dict = None
+
+        state_dict = self.state_dict()
+        data = self.dataset[item]
+        # print(self.worker_id, "data", data, "state", state_dict)
+        return data, state_dict
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:185]:
```suggestion
        return self.dataset[item], state_dict
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -108,6 +108,11 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
             StopIteration: When the epoch is canceled by the user returning -1
         """
         _, (batch, is_last) = next(dataloader_iter)
+        batch, state = sanitize_batch(
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:111]: I thought we would override `_next_data`.

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -170,58 +172,64 @@ def _set_rng_states_on_obj(obj, rng_state_dict: Dict[str, Any]):
             attr.set_state(v)
 
 
+@dataclass
+class IteratorState:
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:176]: Neat !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -91,13 +92,21 @@ def reset(self) -> None:
...
+    def on_run_start(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
         # hook
+        self._dataloader_iter = dataloader_iter
         self.trainer.logger_connector.on_epoch_start()
         self.trainer.call_hook("on_epoch_start")
         self.trainer.call_hook("on_train_epoch_start")
         self.trainer.fit_loop.epoch_progress.increment_started()
 
+        if self.restarting:
+            state = self._dataloader_iter_state
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:104]: Neat ! I was doing state restoration on `on_load_checkpoint`. This is cleaner.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 5]
```diff
@@ -229,6 +238,10 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
 
         self.update_lr_schedulers("epoch", update_plateau_schedulers=True)
 
+        self.batch_progress.current.reset()
+        self.scheduler_progress.current.reset()
+        self.batch_loop.optim_progress.reset_on_epoch()
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:243]: Is it possible for the loop not to access its child loop progress to reset it ?

File: pytorch_lightning/utilities/apply_func.py [hunk 6]
```diff
@@ -66,6 +66,48 @@ def _is_dataclass_instance(obj):
     return dataclasses.is_dataclass(obj) and not isinstance(obj, type)
 
 
+def _remove_empty_collection(coll: Collection):
+    if bool(coll):
+        return coll
+    return None
```
Comment (tchaton) [inline pytorch_lightning/utilities/apply_func.py:72]:
```suggestion
def _remove_empty_collection(collection: Collection):
    if bool(collection):
        return collection
    return None
```

File: pytorch_lightning/utilities/apply_func.py [hunk 7]
```diff
@@ -95,6 +137,11 @@ def apply_to_collection(
     if isinstance(data, dtype) and (wrong_dtype is None or not isinstance(data, wrong_dtype)):
         return function(data, *args, **kwargs)
 
+    # range is a type implemented in C and is afaik the only sequence-like
+    # that does not accept other sequences for construction
+    if isinstance(data, range):
```
Comment (tchaton) [inline pytorch_lightning/utilities/apply_func.py:142]: Super neat !

File: pytorch_lightning/utilities/auto_restart.py [hunk 8]
```diff
@@ -61,17 +68,27 @@ def worker_id(self) -> int:
         return worker_info.id if worker_info else 0
 
     def __iter__(self) -> Iterator[Any]:
-        # the `state dict` was cached as workers were unavailable before
-        # reload it now
-        self._load_cached_state()
+        self._current_iteration = 0
+        print("iter called", self._current_iteration)
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:72]: debug ?

File: pytorch_lightning/utilities/auto_restart.py [hunk 9]
```diff
@@ -61,17 +68,27 @@ def worker_id(self) -> int:
...
+        # the `state dict` was cached as workers were unavailable before.
+        if self._cached_state_dict is not None:  # and self.worker_id in self._cached_state_dict:
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:74]:
```suggestion
        if self._cached_state_dict is not None:
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 10]
```diff
@@ -61,17 +68,27 @@ def worker_id(self) -> int:
...
+            # reload the current state dict
+            self._load_non_random_state(self._cached_state_dict)
 
         i = 0
         sampler_iter = iter(self._sampler)
         while i < self._current_iteration:
+            print("fast forward", i, self._current_iteration)
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:81]: Same here

File: pytorch_lightning/utilities/auto_restart.py [hunk 11]
```diff
@@ -116,13 +139,92 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
+
+    # TODO: only return the state from the latest _get_item()
+    def __getitem__(self, item) -> Tuple[Any, Dict[int, Dict]]:
+        if self._cached_state_dict is not None:
+            if self.worker_id in self._cached_state_dict:
+                set_rng_states(self._cached_state_dict[self.worker_id]["rng_states"])
+            self._cached_state_dict = None
+
+        data = self.dataset[item]
+        state_dict = self._state_dict()
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:196]: Should state dict be captured before fetching the data ?

File: pytorch_lightning/utilities/auto_restart.py [hunk 12]
```diff
@@ -389,7 +493,44 @@ def _sampler_metadata_collate(samples: List, dataset: Dataset, default_collate:
...
+        data = default_collate(samples)
+        metadata = states[-1]
+    else:
+        return default_collate(samples)
+
+    # TODO: change this key name or make a dataclass
+    return {"data": data, AutoRestartBatchKeys.PL_SAMPLERS: metadata}
+
+
+def patch_dataloader_iterator(dataloader: DataLoader, iterator: Iterator):
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:511]: Really neat !
